### PR TITLE
C++: Don't use definitionByReference for data flow

### PR DESCRIPTION
**Pull request by jbj**
The data flow library conflates pointers and objects enough for the `definitionByReference` predicate to be too strict in some cases. It was too permissive in other cases that are now (or will be) handled better by field flow.

See also the change note entry.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll [hunk 1]
```diff
@@ -133,17 +133,35 @@ private module PartialDefinitions {
...
+      argumentType instanceof PointerType and
+      not argumentType.(PointerType).getBaseType().isConst() and
+      (
+        // f(variable)
+        va = argument
+        or
+        // f(&variable)
+        va = argument.(AddressOfExpr).getOperand()
+        or
+        // f(&array[0])
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: should this be generalized to handle multi-dimensional arrays?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: Is it normal to pass a multi-dimensional array as `f(&my3dArray[0][0][0])`? I'd expect people to just write `f(my3dArray)`. What should we do about `f(&my3dArray[0][0])` (two zeros)?

Can you find other common cases that aren't handled right here? I wouldn't be surprised if there are more problems.

File: change-notes/1.22/analysis-cpp.md [hunk 2]
```diff
@@ -24,6 +24,7 @@
 
 - The predicate `Variable.getAnAssignedValue()` now reports assignments to fields resulting from aggregate initialization (` = {...}`).
 - The predicate `TypeMention.toString()` has been simplified to always return the string "`type mention`".  This may improve performance when using `Element.toString()` or its descendants.
+- The `DataFlow::DefinitionByReferenceNode` class now considers `f(x)` to be a definition of `x` when `x` is a variable of pointer type. It no longer considers deep paths such as `f(&x.myField)` to be definitions of `x`. These changes are in line with the user expectations we've observed.
```
Comment (geoffw0) [inline change-notes/1.22/analysis-cpp.md]: Makes sense to me, even if there's a hint of heuristic about this.

**Comment by geoffw0** [comment]
Samate tests job: https://jenkins.internal.semmle.com/job/Security/job/SAMATE/job/SAMATE-cpp/461/

I'm a bit paranoid about the samate job failing because it's only run once per week by default (which is usually plenty), but we're merging a lot of data flow changes this week!

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll (hunk 1 above)
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: I don't think it would be normal to do that, but I wouldn't be surprised if something like `f(&my3dArray[x][y][z])` is significantly more common.

This implicitly covers `FieldAccess` as well - I assume that's intended?

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll [hunk 3]
```diff
@@ -133,17 +133,35 @@ private module PartialDefinitions {
...
+  }
 
-    override predicate partiallyDefines(Variable v) { va = v.getAnAccess() }
+  private predicate referenceArgument(Expr argument, VariableAccess va) {
+    argument = any(Call c).getAnArgument() and
+    exists(Type argumentType |
+      argumentType = argument.getFullyConverted().getType().stripTopLevelSpecifiers()
+    |
+      argumentType instanceof ReferenceType and
+      not argumentType.(ReferenceType).getBaseType().isConst() and
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: What about rvalue references to non-const types?

Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: I believe those are ruled out with this predicate. The argument is restricted to be a `VariableAccess`, and a `VariableAccess` is never an rvalue unless you send it through `std::move` (or cast it yourself, but that should be very rare).

**Comment by geoffw0** [comment]
The Samate test run succeeded.  I'm happy to merge this when we're ready to have it in master.

**Comment by jbj** [comment]
Change note needs to move to 1.23.

**Comment by jbj** [comment]
I've updated this PR to put the change note in the right place.

@rdmarsh2, it's true that these changes are syntactic in nature and incomplete, but it sounds like you agree that the most common cases are covered. How about we leave the more semantically accurate modelling to the IR-based data flow library?

**Comment by jbj** [comment]
Actually, I have another commit coming to this branch to fix the test failure that has started appearing after I rebased. I haven't figured out exactly why the test fails. I think it's a bug that's _revealed_ but not _caused_ by this PR.

**Comment by jbj** [comment]
The tests broke after rebasing this branch on `master`, and I've made a few changes to fix it.
1. The first commit is unchanged. That's the important bit.
2. I dropped the second commit, which was an attempt to simplify definition by reference in data flow. It broke a test and did not seem to give any benefits other than shorter QL code.
3. I added a commit to fix data flow for `ConditionDeclExpr`. The problem is unrelated to this PR and has existed for as long as the data flow library has, but it's convenient to put the fix on this PR since it means I don't have to write a new test for it.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll (hunk 1 above)
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: Looking at this again, I'm not at all sure if `FieldAccess` should be included here.

**Comment by rdmarsh2** [comment]
I'm still unsure about the `FieldAccess` thing, but it looks good otherwise.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll [hunk 4]
```diff
@@ -138,18 +138,42 @@ private module PartialDefinitions {
...
   /**
-   * A partial definition that's a definition by reference (in the sense of the
-   * `definitionByReference` predicate).
+   * A partial definition that's a definition by reference.
    */
   class DefinitionByReference extends PartialDefinition, TReferenceArgument {
     VariableAccess va;
 
-    DefinitionByReference() { definitionByReference(va, definedExpr) }
+    DefinitionByReference() { referenceArgument(va, definedExpr) }
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: It might look strange that `this` is not restricted in this charpred. That's because the full extent of this class includes the charpred of the superclass, which relates `this` to `definedExpr`, and `va` is functionally determined by `definedExpr`.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll (hunk 1 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: Thank you for the reminder. I didn't see the `FieldAccess` comment in your last review.

The `referenceArgument` predicate defines the `DefinitionByReference` class, which is used for two things. The first thing is that it's [used in the `TBlockVar` constructor](https://github.com/Semmle/ql/blob/1bb57daf6fe659195e6323b8e9f8b188f948af33/cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll#L243), but that constructor requires `not v instanceof Field` and so excludes fields. The second thing is that it [determines the charpred of `DefinitionByReferenceNode`](https://github.com/Semmle/ql/blob/067c55adb948f9ef66150775eb24c159d7d9868b/cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll#L198). The `DefinitionByReferenceNode`s that don't assign to fields will have a `FlowVar` associated with them, and there will be flow out of them thanks to [this rule](https://github.com/Semmle/ql/blob/067c55adb948f9ef66150775eb24c159d7d9868b/cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll#L478-L480). Those that assign to fields will not have any flow out of them by default (because they have no `FlowVar`), but they will still be available in case users of the library want to add custom data flow edges out of them with `isAdditionalFlowStep`.

In summary, I think `FieldAccess` should be included here because it makes the library more flexible. You might even argue that arguments that have no `VariableAccess` should be included too, but then we're getting into corner cases that are probably best handled with IR-based data flow.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll (hunk 4 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll]: I've added a commit that turns the above remark into a code comment.
